### PR TITLE
Issue 4526 - sync_repl: when completing an operation in the pending l…

### PR DESCRIPTION
**Pull request by tbordaz**
…ist, it can select the wrong operation

Bug description:
	When an operation complete, it was retrieved in the pending list with
	the address of the Operation structure. In case of POST OP nested operations
	the same address can be reused. So when completing an operation there could be
	a confusion which operation actually completed.

Fix description:
	The fix defines a new operation extension (sync_persist_extension_type).
	This operation extension contains an index (idx_pl) of the op_pl in the
	the pending list.
	And additional safety fix is to dump the pending list in case it becomes large (>10).
	The pending list is dumped with SLAPI_LOG_PLUGIN.

relates: https://github.com/389ds/389-ds-base/issues/4526

Reviewed by:

Platforms tested: F31

### CODE REVIEW

File: ldap/servers/plugins/sync/sync_persist.c
Comment (Firstyear) [inline ldap/servers/plugins/sync/sync_persist.c]: Greater then 10? But then it goes to a loop with slapi_log_plugin? Shouldn't we check if the log level is set as well before we do the dump? or otherwise just log a plain warn?

**Review by Firstyear — CHANGES_REQUESTED** [review]

**Comment by Firstyear** [comment]
I have reviewed this again and I think it's okay. If you don't mind I'd like to give my mind a break and I'll review a second time if thats okay? That way I'm sure i understand it :)

**Comment by tbordaz** [comment]
> I have reviewed this again and I think it's okay. If you don't mind I'd like to give my mind a break and I'll review a second time if thats okay? That way I'm sure i understand it :)

Thank you soooo much @Firstyear for your reviews. I know that I have been so boring with 'pending list' stabilization :)
No problem to wait for a second review.
Just a note regarding the changes. The fix contains basically two aspects:

- use of operation extension to uniquely identify (in the pending list) an operation, because a same operation block is prone to be reused in nested operation (postop operations)
- handling of db_retry that was the most recent issue with pending list (sometime detected with freeipa/dnssec tests). If an update is retried N times, then sync_update_persist_betxn_pre_op is called N times. The main job of sync_update_persist_betxn_pre_op is to add the operation in the pending list and to register its index in the operation extension. The fix uses the operation extension to detect that it is an update retry (op_ext != NULL).

**Review by Firstyear — APPROVED** [review]
Yep, rereading this I'm happy with it :) Thanks @tbordaz! Ack,.

**Pull request merged**
